### PR TITLE
Stage 2 changes for RFC 0009 - data_stream fields

### PR DESCRIPTION
**Pull request by ebeahan**
Adding the `data_stream` fields from [RFC 0009](https://github.com/elastic/ecs/blob/master/rfcs/text/0009-data_stream-fields.md) (#1145) to the experimental artifacts.

### CODE REVIEW

**Review by webmat — COMMENTED** [review]
Initial review, a few things to discuss or adjust :-)

File: experimental/schemas/data_stream.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+---
+- name: data_stream
+  title: Data Stream
+  short: The data_stream fields are part defining the new data stream naming scheme.
```
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: Reworded in both places.

File: experimental/schemas/data_stream.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  description: >
+    The data_stream fields are part defining the new data stream naming scheme.
+    In the new data stream naming scheme the value of the data stream fields combine to the name of the actual data stream in the following manner `{data_stream.type}-{data_stream.dataset}-{data_stream.namespace}`. This means the fields can only contain characters that are valid as part of names of data streams. More details about this can be found in this blog post. TODO: Add link to blog post
```
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: Done.

File: experimental/schemas/data_stream.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    Due to the fact that the values of the `data_stream` fields make up the data stream name, the restrictions on data stream names also apply to values for the `data_stream` fields. As an example, they cannot include \, /, *, ?, ", <, >, |, ` `. Please see the Elasticsearch reference for [restrictions on index/data stream names](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html#indices-create-api-path-params).
```
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: Updated to an asciidoc link.

I'd like to get @ruflin input on this feedback, so I held off on modifying the description for now in my latest changes.

File: experimental/schemas/data_stream.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      short: An overarching type for the data stream.
+
+    - name: dataset
+      level: extended
+      type: constant_keyword
+      example: nginx.access
+      description: >
+        The field can contain anything that makes sense to signify the source of the data.
+        Examples include `nginx.access`, `prometheus`, `endpoint` etc. For data streams that otherwise fit, but that do not have dataset set we use the value "generic" for the dataset value. `event.dataset` should have the same value as `data_stream.dataset`.
+        Data dataset value has the following restrictions:
```
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: I rephrased both here and `.namespace` a bit. Let me know if that conveys these points a bit better now.

File: experimental/schemas/data_stream.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        * Must not contain `-`
+        * No longer than 100 chars
+      short: The field can contain anything that makes sense to signify the source of the data.
+
+    - name: namespace
+      level: extended
+      type: constant_keyword
+      example: logs
```
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: Opted for `production` as the more concrete example.

File: experimental/schemas/data_stream.yml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  fields:
+
+    - name: type
+      level: extended
+      type: constant_keyword
+      example: logs
+      description: >
+        An overarching type for the data stream.
+
+        Currently allowed values include "logs", "metrics". We expect to also add "traces" and "synthetics" in the near future
```
Comment (ruflin) [inline experimental/schemas/data_stream.yml:25]:
```suggestion
        Currently allowed values include "logs", "metrics". We expect to also add "traces" and "synthetics" in the near future.
```

File: experimental/schemas/data_stream.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    - name: type
+      level: extended
+      type: constant_keyword
+      example: logs
+      description: >
+        An overarching type for the data stream.
+        Currently allowed values include "logs", "metrics". We expect to also add "traces" and "synthetics" in the near future
+        Any future values for `data_stream.type` should also adhere to the following restrictions (these are derived from the Elasticsearch index restrictions):
+          * Must not contain `-`
+          * Must not start with `+` or `_`
```
[the remark this answers is not in the document]
Comment (ruflin) [reply experimental/schemas/data_stream.yml:29]: Good point. User should not create their own values.

File: experimental/schemas/data_stream.yml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      type: constant_keyword
+      example: nginx.access
+      description: >
+        The field can contain anything that makes sense to signify the source of the data.
+
+        Examples include `nginx.access`, `prometheus`, `endpoint` etc. For data streams that otherwise fit, but that
+        do not have dataset set we use the value "generic" for the dataset value. `event.dataset` should have the
+        same value as `data_stream.dataset`.
+
+        Beyond the Elasticsearch index naming criteria noted above, the `dataset` value has additional restrictions:
```
Comment (ruflin) [inline experimental/schemas/data_stream.yml:43]: I'm torn if we should talk about index naming criteria or data stream naming criteria.

File: experimental/schemas/data_stream.yml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    - name: namespace
+      level: extended
+      type: constant_keyword
+      example: production
+      description: >
+        A user defined namespace. Namespaces are useful to allow grouping of data.
+
+        Many users already organize their indices this way, and the data stream naming scheme now provides this
+        best practice as a default. Many users will populate this field with `default`.
```
Comment (ruflin) [inline experimental/schemas/data_stream.yml:56]:
```suggestion
        best practice as a default. Many users will populate this field with `default`. If not value is used it falls back to `default`.
```

**Review by ruflin — DISMISSED** [review]
LGTM. I left a few minor comments / suggestions.

File: experimental/schemas/data_stream.yml (hunk 8 above)
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: I opted for data stream here.

File: experimental/schemas/data_stream.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: Converted this limitation section to a comment, as proposed.

File: experimental/schemas/data_stream.yml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      level: extended
+      type: constant_keyword
+      example: logs
+      # Any future values for `data_stream.type` should also adhere to the following restrictions (these are derived from the Elasticsearch index restrictions):
+      # * Must not contain `-`
+      # * Must not start with `+` or `_`
+      description: >
+        An overarching type for the data stream.
+
+        Currently allowed values include "logs", "metrics". We expect to also add "traces" and "synthetics" in the near future.
```
Comment (webmat) [inline experimental/schemas/data_stream.yml:28]: I suggest being just a tad more forceful.

```suggestion
        Currently allowed values are "logs", "metrics". We expect to also add "traces" and "synthetics" in the near future.
```

File: experimental/schemas/data_stream.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ebeahan) [reply experimental/schemas/data_stream.yml]: I reworked this section a bit, incorporating some of @webmat's feedback to better capture that there's a relationship between a data stream's name and the name of the data stream's backing indices, which enforces similar naming limitations.

Let me know if it's still a bit off.

**Review by webmat — APPROVED** [review]
Yeah I like the new wording to explain ds naming vs index name restrictions 👍

**Pull request merged**